experimental/ast/decl_def.go

```diff
@@ -27,12 +45,7 @@ import "github.com/bufbuild/protocompile/internal/arena"
 //
 // Generally, you should not need to work with DeclDef directly; instead, use the As* methods
 // to access the correct concrete syntax production a DeclDef represents.
-type DeclDef struct {
-	withContext
-
-	ptr arena.Untyped
-	raw *rawDeclDef
-}
+type DeclDef struct{ declImpl[rawDeclDef] }
```

Comment:
I like how this is turning out with these latest changes!

Possibly a controversial suggestion (apologies in advance):

I observe that `DeclDef` functions a lot like a `DefAny`. What if we named it `DeclDefAny` and then renamed the others to have a `DeclDef` name prefix instead of just `Def` (e.g. `DeclDefMessage`...).

I know the logic behind the way these vary is different than the other `Decl*`, `Expr*`, and `Type*`. But it does it _need_ to be different, in terms of API cohesion? Seems like we could name things and represent them in a very similar way, including an `AsAny() DeclDefAny` method on the concrete def types, instead of an exported field. That would provide better API symmetry/consistency across all of these things.

WDYT?

Reply:
Interesting. I do think `DeclDefMessage` is a little uh... wordy? (but it's fun to say) We *could* have something like
```
type DeclDef struct {
  Def DefAny
}
```
Not unlike how we have `TypePath`.

There is one problem, which is that mutating a field in a Def does not update it when you call AsAny, and making that happen kind of defeats the purpose of the Def types at all, in my mind. Also, unlike all of the other `Kind*` types, `DeclDefAny` would allow you to extract a richer state than that suggested by the concrete types (which is necessary for legalization, given how parsing for defs works: if I've decided something is a message, and for some reason the user wrote `message Foo = 43;` (which we will parse as a generic def), I need to complain about the `= 43`.